### ISSUE TITLE
Providers with onInit logic cannot have injected dependencies overwritten

### ISSUE THREAD

**Issue by lavigneer** [Urigo/graphql-modules#1115]
I'm running into an issue with overwriting dependencies of providers that use the onInit hook.

Say we have the following:

```
const CONN_TOKEN = 'CONN_TOKEN'

@Injectable()
class ConnectionProvider {
    call(method, params) {
        // some logic that send a request to the connection
    }
}

const ConnectionModule = {
    providers: [
        {
            provide: CONN_TOKEN,
            useClass: () => ConnectionProvider
        }
    ]
}

@Injectable()
class ServiceProvider {
    constructor(@Inject(CONN_TOKEN) connection) {
        this.connection = connection
    }

    onInit() {
        this.connection.call('someMethod', 'someParams')
    }
}

const ServiceModule = {
    providers: [ServiceProvider],
    imports: [ConnectionModule]
}
```

And we have a test where we want to overwrite the `ConnectionProvider` to `ServiceModule`'s by doing the following:

```
const mockModule = ServiceModule.mock({})

mockModule.injector.provide({
    provide: CONN_TOKEN,
    useValue: {
        call: (method, params) => {
            // different call logic
        }
    }
})

// some test logic that uses mockModule 
```

The problem here is that the `onInit` function of `ServiceProvider` gets called before we can call `mockModule.injector.provide` so it's using an instance of the `ConnectionProvider` rather than the supplied overwrite of that provider.

Is there a way to intercept the injector to apply these overwrites before it triggers the `onInit` hook?

**Comment by kamilkisiela** [comment]
We released a new major version (v1.0) of GraphQL Modules. Here's the [migration guide](https://graphql-modules.com/docs/next/recipes/migration).

The new version solves the issue by not enforcing dependencies between modules and easier DI handling on application level, so I'm closing the thread.

Please try to upgrade. If you can't and still want to use the legacy version, please open a new issue.
